File: compressor.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileOutputStream;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;


public class compressor {

    static class TxtReader {
        public static String readTextFile(String filename) throws IOException {
            StringBuilder stringBuilder = new StringBuilder();
            try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
                String line;
                while ((line = br.readLine()) != null) {
                    stringBuilder.append(line);
                    stringBuilder.append('\n');
                }
            }
            return stringBuilder.toString();
        }
    }

    static class Compressor {
        public byte[] compress(String text) {
            Map<String, Integer> wordToNumber = new HashMap<>();
            int wordCount = 0;
            StringBuilder compressedText = new StringBuilder();

            String[] words = text.split("\\s+");

            for (String word : words) {
                if (!wordToNumber.containsKey(word)) {
                    wordToNumber.put(word, ++wordCount);
                }

                int wordNumber = wordToNumber.get(word);

                // Add the word number to the compressed text
                compressedText.append((char) wordNumber);
            }

            // Convert compressed text to byte array
            byte[] compressedArray = new byte[compressedText.length()];
            for (int i = 0; i < compressedText.length(); i++) {
                compressedArray[i] = (byte) compressedText.charAt(i);
            }

            return compressedArray;
        }
    }

    static class CompressedFileWriter {
        public static void writeCompressedFile(byte[] compressedData, String filename) throws IOException {
            try (FileOutputStream result = new FileOutputStream(filename + ".sc")) {
                result.write(compressedData);
            }
        }
    }

    static class Decompressor {
        public static String decompress(byte[] compressedData) {
            Map<Integer, String> numberToWord = new HashMap<>();
            int wordCount = 0;

            StringBuilder decompressedText = new StringBuilder();
            int i = 0;
            while (i < compressedData.length) {
                int count = compressedData[i++];
                int wordNumber = compressedData[i++];
                String word = numberToWord.get(wordNumber);
                if (word == null) {
                    word = String.valueOf(++wordCount);
                    numberToWord.put(wordNumber, word);
                }
                for (int j = 0; j < count; j++) {
                    decompressedText.append(word);
                }
            }
            return decompressedText.toString();
        }
    }
    public class DecompressedFileWriter {
        public static void writeDecompressedFile(String decompressedText, String filename) throws IOException {
            try (BufferedWriter decompressed_result = new BufferedWriter(new FileWriter(filename))) {
                decompressed_result.write(decompressedText);
            }
        }
    }

    static class TextComparer {
        public static boolean isSame(String originalText, String decompressedText) {
            return originalText.equals(decompressedText);
        }
    }

    public static void main(String[] args) {
        try {
            String originalText = TxtReader.readTextFile("/Users/saadatibakova/Desktop/mytext.txt");
            Compressor compressor = new Compressor();
            byte[] compressedData = compressor.compress(originalText);
            CompressedFileWriter.writeCompressedFile(compressedData, "compressed");
            String decompressedText = Decompressor.decompress(compressedData);
            boolean isSame = TextComparer.isSame(originalText, decompressedText);
            System.out.println("Is decompressed text same as original? " + isSame);
            DecompressedFileWriter.writeDecompressedFile(decompressedText, "decompressed.txt");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: Compressor.py
class Compressor:
    def __init__(self):
        self.word_to_num = {}
        self.num_to_word = {}
        self.next_id = 1

    def compress(self, text):
        words = text.split()
        compressed_words = []

        for word in words:
            if word not in self.word_to_num:
                self.word_to_num[word] = self.next_id
                self.num_to_word[self.next_id] = word
                self.next_id += 1

            compressed_words.append(str(self.word_to_num[word]))

        return ' '.join(compressed_words)

    def decompress(self, compressed_text):
        nums = compressed_text.split()
        decompressed_words = []

        for num in nums:
            word_id = int(num)
            if word_id in self.num_to_word:
                decompressed_words.append(self.num_to_word[word_id])
            else:
                raise ValueError(f"ID {word_id} not found in the decompression dictionary.")

        return ' '.join(decompressed_words)

    def compress_file(self, input_file, output_file):
        with open(input_file, 'r') as f_in:
            text = f_in.read()
        compressed_text = self.compress(text)
        with open(output_file, 'w') as f_out:
            f_out.write(compressed_text)
        print(f"File '{input_file}' compressed to '{output_file}'")

    def decompress_file(self, input_file, output_file):
        with open(input_file, 'r') as f_in:
            compressed_text = f_in.read()
        decompressed_text = self.decompress(compressed_text)
        with open(output_file, 'w') as f_out:
            f_out.write(decompressed_text)
        print(f"File '{input_file}' decompressed to '{output_file}'")

# Example usage:
if __name__ == "__main__":
    compressor = Compressor()
    
    # Compress a file
    compressor.compress_file('/Users/saadatibakova/Desktop/mytext.txt', 'compressed.txt')
    
    # Decompress the file
    compressor.decompress_file('compressed.txt', 'output.txt')


